04cpp3/sec02Templates.md

```diff
@@ -0,0 +1,351 @@
+
+Templates in C++ come in two main kinds:
+- Function Templates
+- Class Templates
+
+When a class or function template is used to instantiate an concrete class or function using a specific type, a new class or function definition is created for each type with which the template is instantiated. So unlike our inheritance based run-time polymorphism, where we have one function which can take multiple classes with overlapping definitions (defined by inheritance from a base class), now we use one piece of code two generate multiple separate functions (or classes), each of which accepts a different type. This is why the compiler must know all the types which are to be used in templated functions at compile time. This is sometimes known as "static polymorphism". 
+
+## Using Templates with Classes
+
+You have already been using templates in this course: `vector<>` is an example of a _class template_. The `vector` class template is defined the type of the elements of the vector left as a template argument: you provide this argument in the angle brackets (`<>`) when you declare your vector object. This then tells the compiler what kind of data your vector will hold, so that it can properly handle it (allocating memory, iterating through elements, type checking for elements which are used in the code etc.). 
+
+**N.B.** A note on terminology:
+- A **class template** is not a class, it is a template which can be used to generate the definition of a class once the template arguments have been filled. For example, `vector<>` is a class template which can be used to generate vector classes which hold different types of data, but you cannot create a vector object with no type in the angle brackets. 
+- A **template class** is a class that has been created from a template. For example `vector<int>` is a template class that has been generated from the `vector` class template by providing the type `int` as a template argument. `vector<int>` is an entirely separate class from `vector<double>`.
+
+The other containers in the standard library are also class templates, which require you to provide template arguments to instantiate concrete classes. Note that some may take more than one template argument, for example `map<>` takes two types: one for the keys and one for the values. Type for a map from strings to integers would be `map<string, int>`. 
+
+We can define a class template using the following syntax:
+```cpp
+template<typename T>
+class myClassTemplate
+{
+    public:
+    myClassTemplate(T value)
+    {
+        myMemberT = value;
+    }
+
+    private:
+    T myMemberT;
+};
+```
+- `T` is the template parameter, and the `typename` keyword tells us that `T` must denote a type. (You can equivalently use the `class` keyword.)
```

Comment:
Maybe mention that it doesn't need to be called `T`, and in fact probably should be called something more readable where relevant?